### PR TITLE
Regression of other extensions for Visual Studio (#5316)

### LINKED ISSUE #5316: [Regression] Two dots in lib filename
I use conan 1.16.0 on windows 10 and I have a package that has a lib file named `GTEngine.v12.lib`. If I run `conan install` which uses the `visual_studio_multi` generator, it adds the lib filename _without extension_ in `conanbuildinfo_debug_x64_v120.props` (I use x64/msvc2013 which is irrelevant). Version 1.15.0 keeps the `.lib` suffix in the property sheet, but 1.16.0 somehow removes it which results in a build error.

Thanks in advance

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

### PR DESCRIPTION
**Pull request by uilianries**
- Regrets #5254 to fix #5316
- Add new tests to validate any possible regression

Changelog: Fix: Accept only .lib and .dll as Visual extensions (#5316)
Docs: Omit

- [x] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

@tags: slow

### CODE REVIEW

File: conans/test/unittests/client/generators/visual_studio_multi_test.py [hunk 1]
```diff
@@ -86,3 +86,50 @@ def valid_xml_test(self, use_toolset):
...
+    def addional_dependencies_test(self):
+
+        def validate_additional_dependencies(libname, additional_dep):
+            tempdir = temp_folder()
+            with chdir(tempdir):
+                settings = Settings.loads(default_settings_yml)
+                settings.os = "Windows"
+                settings.compiler = "Visual Studio"
+                settings.compiler.version = "11"
+                settings.compiler.runtime = "MD"
```
Comment (jgsogo) [inline conans/test/unittests/client/generators/visual_studio_multi_test.py]: Move these settings with the others (around line 108), and remove duplicated ones (`compiler.version`)

File: conans/client/generators/visualstudio.py [hunk 2]
```diff
@@ -65,13 +65,17 @@ def filename(self):
         return BUILD_INFO_VISUAL_STUDIO
 
     def _format_properties(self, build_info, condition):
+        def has_valid_ext(lib):
+            ext = os.path.splitext(lib)[1]
+            return ext if ext in (".lib", ".dll") else None
```
[the remark this answers is not in the document]
Comment (uilianries) [reply conans/client/generators/visualstudio.py]: it's a valid extension no? The previous feature enabled custom extensions.

File: conans/client/build/visual_environment.py [hunk 3]
```diff
@@ -78,7 +78,8 @@ def _get_link_list(self):
         # FIXME: Conan 2.0. The libs are being added twice to visual_studio
         # one in the conanbuildinfo.props, and the other in the env-vars
         def format_lib(lib):
-            return lib if os.path.splitext(lib)[1] else '%s.lib' % lib
+            ext = os.path.splitext(lib)[1]
+            return lib if ext in (".lib", ".dll") else '%s.lib' % lib
```
[the remark this answers is not in the document]
Comment (uilianries) [reply conans/client/build/visual_environment.py]: ditto

File: conans/test/unittests/client/generators/visual_studio_multi_test.py (hunk 1 above)
Comment (uilianries) [reply conans/test/unittests/client/generators/visual_studio_multi_test.py]: done

File: conans/client/generators/visualstudio.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (uilianries) [reply conans/client/generators/visualstudio.py]: done, applied your suggestion.

**Comment by uilianries** [comment]
@memsharded could you confirm if we should accept .dll as extension or just .lib is fine?

**Comment by memsharded** [comment]
Hi @uilianries 

Please do this PR to release/1.16.1 branch, it is a regression, should be fixed asap. Many thanks!

**Comment by uilianries** [comment]
@memsharded done

**Comment by uilianries** [comment]
I think is better recreating a new PR, because I started this one from develop branch.

**Comment by uilianries** [comment]
forced push because I moved the base branch to release/1.16.1 and cherry-pick all 4 commits

**Comment by memsharded** [comment]
Added tags=slow, and relaunched, as this will break the new test introduced in #5254. Unfortunately, the ``mylib.a`` is also a valid extension, used by some users, and it totally work, that test proves it.

**Comment by memsharded** [comment]
My bad, it is not broken, but the test is not running, because the filename is not adding _test. I will fix it asap.

**Comment by memsharded** [comment]
The CI is broken, this PR needs to be relaunched after fixing it

**Comment by memsharded** [comment]
I have pushed changes, adding ".a" to the set of valid extensions for the generator.

**Comment by jgsogo** [comment]
I think that **if we are going to change the extension list (previously it was only `.lib`)** it should be the same we are using in `collect_libs`, somehow this two methods (collect and generate) are related: https://github.com/conan-io/conan/blob/551ddac6425341a861258617a2fbd09faf4a71ba/conans/client/tools/files.py#L324

`collect_libs` is removing some extensions, and generators are adding them again. Given this, I will limit the list to `.lib` and `.a`.

**With this change, we are breaking users with library names like `mylib.a.lib`** (using `cpp_info.libs = ["mylib.a"]`), although we are helping some users with custom library extensions:
 * `.a`: `collect_libs` is working for them, now they won't get `.a.lib` in the generator. Was that change (#5254) motivated by any issue? Nevertheless, I agree with it even though we may break someone else (no one has complained yet).
 * `.dll`: `collect_libs` is not working for them, still not working, but they can start to use the generator. <-- who uses `.dll` as a custom suffix for import libraries? 🤷‍♂ Please, reconsider.

Any other custom exception will still be broken because the generator will add `.lib` to it although `collect_libs` won't trim it ([comment here](https://github.com/conan-io/conan/issues/5303#issuecomment-499475476)) .

**Comment by uilianries** [comment]
> I think that if we are going to change the extension list (previously it was only .lib) it should be the same we are using in collect_libs

I think it's a good idea. We could re-use the same code from there.

**Comment by memsharded** [comment]
> collect_libs is removing some extensions, and generators are adding them again. Given this, I will limit the list to .lib and .a.

Not true, ``collect_libs`` is removing all extensions. For those extensions, it is also removing the ``lib`` prefix.

**Comment by jgsogo** [comment]
> > collect_libs is removing some extensions, and generators are adding them again. Given this, I will limit the list to .lib and .a.
> 
> Not true, `collect_libs` is removing all extensions (the last one only in case of several extensions). For those extensions, it is also removing the `lib` prefix.

Not exactly, it adds the file only if it has one of those extensions, and removes the `lib` prefix for all but those with suffix `.lib`. This is the current implementation:

```python
if ext in (".so", ".lib", ".a", ".dylib", ".bc"): 
    if ext != ".lib" and name.startswith("lib"):
        name = name[3:]
    # ... do things
    result.append(name)
```

File: conans/client/generators/visualstudio.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 
 from conans.model import Generator
 from conans.paths import BUILD_INFO_VISUAL_STUDIO
+from conans.client.tools.files import _valid_lib_extensions
```
Comment (memsharded) [inline conans/client/generators/visualstudio.py]: Not nice to have to import a private thing from another module, better make it public

Comment (uilianries) [reply conans/client/generators/visualstudio.py]: done!

**Comment by uilianries** [comment]
OSX failed due server error: https://conan-ci.jfrog.info/blue/organizations/jenkins/ConanTestSuite/detail/PR-5319/15/pipeline/150#step-177-log-1354

need to relaunch CI

**Comment by uilianries** [comment]
> About exposing it to conans.tools.valid_lib_extensions, right now I don't find it useful for the user (and it is not documented).

What about using as protected?

File: conans/tools.py [hunk 5]
```diff
@@ -94,6 +94,7 @@ def get(*args, **kwargs):
 which = tools_files.which
 unix2dos = tools_files.unix2dos
 dos2unix = tools_files.dos2unix
+valid_lib_extensions = tools_files.valid_lib_extensions
```
Comment (memsharded) [inline conans/tools.py]: No, certainly we don't want to expose it for usage by recipes.
By default we don't expose anything to tools, unless it has been explicitly requested and approved.

File: conans/client/tools/files.py [hunk 6]
```diff
@@ -304,7 +304,7 @@ def _path_equals(path1, path2):
     return path1 == path2
 
 
-def _valid_lib_extensions():
+def valid_lib_extensions():
```
Comment (memsharded) [inline conans/client/tools/files.py]: Agree with @jgsogo, better declare just a tuple, not a function

File: conans/tools.py (hunk 5 above)
Comment (uilianries) [reply conans/tools.py]: So I need to import as "private"

File: conans/client/tools/files.py (hunk 6 above)
Comment (uilianries) [reply conans/client/tools/files.py]: done!

File: conans/test/unittests/client/generators/visual_studio_test.py [hunk 7]
```diff
@@ -104,3 +104,29 @@ def multi_config_test(self):
...
+            cpp_info = CppInfo("dummy_root_folder1")
+            cpp_info.libs = [lib]
+            conanfile.deps_cpp_info.update(cpp_info, ref.name)
+            generator = VisualStudioGenerator(conanfile)
+            self.assertIn("<AdditionalDependencies>"
+                          "{};%(AdditionalDependencies)"
+                          "</AdditionalDependencies>".format(additional_dep), generator.content)
+
+        # regular
+        validate_additional_dependencies("foobar", "foobar.lib")
```
Comment (SSE4) [inline conans/test/unittests/client/generators/visual_studio_test.py]: may we add two tests here just to be sure:
1. libiconv use-case : `libiconv.dll.lib`
2. ffmpeg use-case: `libavcodec.a`

Comment (uilianries) [reply conans/test/unittests/client/generators/visual_studio_test.py]: done

File: conans/tools.py (hunk 5 above)
Comment (memsharded) [reply conans/tools.py]: No, just do not import it from tools. It is not a tool, it is for internal usage only.

**Pull request merged**